ckanext/canada/helpers.py

```diff
@@ -758,3 +766,98 @@ def fgp_viewer_url(package):
             openmap_uri = 'openmap'
 
         return h.adv_search_url() + '/' + openmap_uri + '/' + package.get('id')
+
+
+def date_field(field, pkg):
+    if pkg.get(field) and ' ' in pkg.get(field):
+        return pkg.get(field).split(' ')[0]
+    return pkg.get(field, None)
+
+
+def split_piped_bilingual_field(field_text, client_lang):
+    if field_text is not None and ' | ' in field_text:
+        return field_text.split(' | ')[1 if client_lang == 'fr' else 0]
+    return field_text
+
+
+def search_filter_pill(search_field, search_extras):
+    html_output = ''
+
+    if search_field == 'portal_type':
+
+        # Custom dataset type labels (suggested datasets discontinued - 2024-05-09)
+        preset_choices = [{'value': 'dataset', 'label': _('Open Data')},
+                          {'value': 'info', 'label': _('Open Information')},]
+
+        # Add PD types
+        for pd_type in h.recombinant_get_types():
+            preset_choices.append({'value': pd_type, 'label': _(h.recombinant_get_chromo(pd_type).get('title'))})
+
+    elif search_field == 'status':
+
+        preset_choices = [{'value': 'department_contacted',
+                           'label': _('Request sent to data owner - awaiting response')}]
+
+    elif search_field == 'ready_to_publish':
+
+        preset_choices = [{'value': 'true', 'label': _('Pending')},
+                          {'value': 'false', 'label': _('Draft')},]
+
+    else:
+
+        preset_choices = (h.scheming_get_preset('canada_' + search_field) or {}).get('choices', [])
+
+        if search_field == 'collection':
+
+            collection_choices = preset_choices
+            preset_choices = [{'value': 'pd', 'label': _('Proactive Publication')}]
+
+            for collection_type in collection_choices:
+                preset_choices.append(collection_type)
+
+            # Add PD types
+            for pd_type in h.recombinant_get_types():
+                preset_choices.append({'value': pd_type, 'label': _(h.recombinant_get_chromo(pd_type).get('title'))})
+
+    def remove_filter_button_html(_field, _value, s_extras, _choices):
+
+        extrs = s_extras.copy() if s_extras else {}
+        extrs.update(request.view_args)
+        link = h.remove_url_param(_field, value=_value, extras=extrs)
+
+        html = '<a href="{link_href}" class="btn btn-info btn-xs mrgn-lft-sm mrgn-bttm-sm facet-pill" title="{link_title}">'.format(
+                    link_href=link,
```

Comment:
in general I avoid building HTML in helpers because it's easy to miss escaping a value and create opportunities for XSS or broken HTML. Consider returning the data you need from the helper and using a template to render the HTML elements instead.

Also repeatedly appending to is an O(n²) operation (although I think newer pythons have a hack to speed them up a bit)

Reply:
@wardi doing a list of tuples now in the helper for the return. And outputting the html in jinja. Are you able to review again? Thanks!